docs/journal/02-exploring-recent-events.md

```diff
@@ -1,7 +1,122 @@
+
+### Stream Analytics + SQL
+
+[Azure Stream Analytics] is a real-time stream processing service. It uses an SQL-like language to process incoming data. Unlike traditional SQL, Stream Analytics runs queries over an event stream, making it easy to analyze events within a temporal window.     
+
+In this approach, we use Stream Analytics to aggregate the event data in real time and store the results in SQL. The aggregated result is small enough that SQL can handle the load.
+
+![Stream analytics diagram](./media/03-ad-hoc-exploration/stream-analytics.png)
+
+
+**Advantages:**
+
+- Easy to configure. Stream Analytics uses a declarative model to specify the input and output streams, and the transformations to be performed by the processing.
+- Easy to manage. Stream Analytics is an Azure service, so you don't need to provision or maintain VMs. 
+- Stream Analytics can pull in *reference data* from blob storage. Reference data is static or infrequently changing data. For example, in our Fabrikam scenario, we use this look up the building number from the device ID.   
+- Stream Analytics is integrated with Event Hubs, which we picked for event ingestion. (See [Journal entry 1][journal1].)
+
+**Concerns: **
+
+- You have to configure the processing ahead of time. This limits the idea of "ad hoc" exploration somewhat, because the user is running queries against the aggregated results, not the raw event data.
+
+
+### Elasticsearch
+
+[Elasticsearch] is an open-source analytics engine. It uses a distributed store and indexes as it writes. You can use a REST API to query, or use the [Kibana] web front end.
+
+In this approach, a custom event processor fetches the data from Event Hubs and writes it to Elasticsearch. To get enough throughput, you might need to use batch writes. 
+
+![Elasticsearch diagram](./media/03-ad-hoc-exploration/elasticsearch.png)
+
+
+**Advantages:**
+
+- **Ease of use**. The [Kibana] dashboard provides an easy way to query and visualize the data.
+- **Performance**. Elasticsearch can perform fast queries against large data sets. 
+- **Scalability**. Elasticsearch uses a distributed document store, so it scales horizontally. 
+- **Data fidelity**. This approach lets you query over the raw event data, rather than aggregating it first. 
+
+**Concerns:**
+
+- **Maintenance**. Elasticsearch is not an Azure service, so you must set up VMs, install and configure Elasticsearch, monitor the health of the cluster, and so forth. ([This blog post][elasticsearch-on-azure] has a nice write up of the steps.)
+- **Security**. Elasticsearch does not provide any security out of the box. You must run it inside a virtual network. 
+- **Not relational**. Elasticsearch is not relational and cannot use reference data (that is, data from another store). To add extra information such as the building number, the event processor must add it to the event data.
```

Comment:
Another approach I've heard is to have a process that pulls data from Elasticsearch, adds the reference data, and then updates Elasticsearch.